#! /bin/sh -e

# All lines beginning with `# DPATCH:' are a description of the patch.
# DP: boehm-gc:
# DP: the thread suspend handler has to save all registers
# DP: re-enable MPROTECT_VDB, it should work, otherwise it's probably a kernel bug
# DP: change STACKBOTTOM to LINUX_STACKBOTTOM so it works with 2.6 kernel

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        #cd ${dir}gcc && autoconf2.59
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        #rm ${dir}gcc/configure
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

--- boehm-gc/include/private/gcconfig.h.orig	2006-08-29 14:06:13.870718000 +0200
+++ boehm-gc/include/private/gcconfig.h	2006-08-29 14:10:28.090718000 +0200
@@ -665,10 +665,10 @@
 #   endif
 #   ifdef LINUX
 #       define OS_TYPE "LINUX"
-#       define STACKBOTTOM ((ptr_t)0xf0000000)
+#       define LINUX_STACKBOTTOM
 #       define USE_GENERIC_PUSH_REGS
 		/* We never got around to the assembly version. */
-/* #       define MPROTECT_VDB - Reported to not work  9/17/01 */
+#       define MPROTECT_VDB
 #       ifdef __ELF__
 #            define DYNAMIC_LOADING
 #	     include <features.h>
--- boehm-gc/pthread_stop_world.c.orig	2006-08-29 14:06:13.870718000 +0200
+++ boehm-gc/pthread_stop_world.c	2006-08-29 14:07:18.810718000 +0200
@@ -124,7 +124,7 @@
 
 void GC_suspend_handler_inner(ptr_t sig_arg);
 
-#if defined(IA64) || defined(HP_PA)
+#if defined(IA64) || defined(HP_PA) || defined(M68K)
 extern void GC_with_callee_saves_pushed();
 
 void GC_suspend_handler(int sig)
